cpp/dolfinx/common/utils.cpp

```diff
@@ -24,3 +29,137 @@ std::string dolfinx::common::indent(std::string block)
   return s.str();
 }
 //-----------------------------------------------------------------------------
+std::pair<std::shared_ptr<const IndexMap>, std::vector<std::int64_t>>
+common::compress_index_map(std::shared_ptr<const IndexMap> map,
+                           const xtl::span<const std::int32_t>& indices)
+{
+  const std::vector<std::int64_t>& ghosts = map->ghosts();
+
+  std::vector<int> ghost_owners = map->ghost_owner_rank();
+
+  // Commmunicate ghost indices
+  MPI_Comm reverse_comm
+      = map->comm(dolfinx::common::IndexMap::Direction::reverse);
+  MPI_Comm forward_comm
+      = map->comm(dolfinx::common::IndexMap::Direction::forward);
+  auto [source_ranks, dest_ranks] = dolfinx::MPI::neighbors(forward_comm);
+
+  // Create inverse map of src ranks
+  std::unordered_map<std::int32_t, std::int32_t> rank_glob_to_loc;
+  for (std::size_t i = 0; i < source_ranks.size(); ++i)
+    rank_glob_to_loc.insert({source_ranks[i], i});
+  // Split indices into local and ghost indices
+  const std::int32_t local_size = map->size_local();
+  std::vector<std::int32_t> local_indices;
+  std::vector<std::int32_t> ghost_indices;
+  std::vector<std::int32_t> num_ghosts(source_ranks.size());
+
+  for (std::size_t i = 0; i < indices.size(); ++i)
+  {
+    if (indices[i] < local_size)
+      local_indices.push_back(indices[i]);
```

Comment:
Could we preallocate the vector? 
Either reserving `local_indices(indices.size())` or by counting ?